File: EduSphere_backend/analytics-service/src/main/java/com/edusphere/analytics/service/AnalyticsService.java
package com.edusphere.analytics.service;

import com.edusphere.analytics.dto.request.GenerateReportRequest;
import com.edusphere.analytics.dto.request.ProgressUpdateRequest;
import com.edusphere.analytics.dto.response.KpiResponse;
import com.edusphere.analytics.dto.response.ReportResponse;
import com.edusphere.analytics.dto.response.StudentProgressResponse;

import java.util.List;
import java.util.UUID;

public interface AnalyticsService {

    KpiResponse getKpis(String courseId, String deptId);

    ReportResponse generateReport(GenerateReportRequest request, UUID generatedBy);

    List<ReportResponse> listReports(UUID requesterId);

    ReportResponse getReport(UUID reportId);

    void updateProgress(ProgressUpdateRequest request);

    List<StudentProgressResponse> getStudentProgress(UUID studentId);

    List<StudentProgressResponse> getCourseProgress(UUID courseId);

    List<StudentProgressResponse> getStudentCourseProgress(UUID studentId, UUID courseId);
}

File: EduSphere_frontend/src/services/analyticsService.js
import api from './api'

export const analyticsService = {
  // KPIs — ADMIN/INSTRUCTOR/COORDINATOR only
  getKpis: (params) => api.get('/analytics/kpis', { params }),

  // Student progress analytics
  getStudentProgress: (studentId) => api.get(`/analytics/progress/student/${studentId}`),

  // Course progress analytics
  getCourseProgress: (courseId) => api.get(`/analytics/progress/course/${courseId}`),

  // Student progress within a specific course
  getStudentCourseProgress: (studentId, courseId) =>
    api.get(`/analytics/progress/student/${studentId}/course/${courseId}`),
}
